### PR TITLE
Add a Scaleway IP module

### PR DESCRIPTION
**Pull request by remyleone**
##### SUMMARY

This PR adds a management module for Scaleway reserved IP.

Here is the list of targeted features:

- Create a new reserved IP address
- Attach an ip address to a server
- Update an IP address mapping (move)
- Remove an IP address

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

- scaleway_ip

##### ANSIBLE VERSION

```
ansible 2.8.0.dev0 (scaleway_ip 7f0f872b2b) last updated 2018/09/03 17:14:39 (GMT +200)
  config file = None
  configured module search path = [u'/Users/sieben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sieben/workspace/ansible/lib/ansible
  executable location = /Users/sieben/workspace/ansible/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

**Comment by remyleone** [comment]
@g1franc I started working on your dynamic IP management module. I will let you know once it is ready for review.

**Comment by g1franc** [comment]
Tested and working

**Comment by remyleone** [comment]
@Spredzy @maxamillion Could you merge it?

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py [hunk 1]
```diff
@@ -29,6 +29,13 @@
 
 options:
 
+  dynamic_ip_required:
+    description:
+    - Enable a public IP at server creation
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: I propose to use:
- an hardcoded IP, for example: `163.172.152.10` (or scaleway address ID?) (IP must exist and be available in the pool of reserved IPs)
- attached (create an IP which will be destroyed when the node will be destroyed)
- absent (don't use any public IP)

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+      - ams1
+      - EMEA-NL-EVS
+      - par1
+      - EMEA-FR-PAR1
+
+  id:
+    description:
+    - id of the IP
+
+  server:
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: Should not that be handled by the `scaleway_compute` module ?

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    changed = True
+    if api.module.check_mode:
+        return changed, {"status": "IP attributes would be changed."}
+
+    ip_patch_response = api.patch(path="ips/%s" % target_ip["id"],
+                                  data=patch_payload)
+
+    if not ip_patch_response.ok:
+        api.module.fail_json(msg='Error during IP attributes patching: [{0}: {1}]'.format(
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: s/patching/update/ ?

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        "reverse": module.params["reverse"],
+        "id": module.params["id"],
+        "server": module.params["server"]
+    }
+
+    region = module.params["region"]
+    module.params['api_url'] = SCALEWAY_LOCATION[region]["api_endpoint"]
+
+    api = Scaleway(module=module)
+    changed, summary = state_strategy[module.params["state"]](api=api, wished_ip=wished_ip)
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]:
```
if module.params["state"] == absent:
    absent_strategy(pi=api, wished_ip=wished_ip)
else:
    present_strategy(pi=api, wished_ip=wished_ip)
```
could be more readable.

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    module.exit_json(changed=changed, scaleway_ip=summary)
+
+
+def main():
+    argument_spec = scaleway_argument_spec()
+    argument_spec.update(dict(
+        state=dict(default='present', choices=['absent', 'present']),
+        organization=dict(),
+        server=dict(),
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: I would let the `scaleway_compute` handles IP assignation to server.

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    required: true
+    choices:
+      - ams1
+      - EMEA-NL-EVS
+      - par1
+      - EMEA-FR-PAR1
+
+  id:
+    description:
+    - id of the IP
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: It doesn't make sense to use `id` parameter with `state=present`. Currently when `state=present` is used and `id` is specified, a new IP with another `id` is created.

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+  reverse:
+    description:
+    - Reverse to assign to the IP
+'''
+
+EXAMPLES = '''
+  - name: Create an IP
+    scaleway_ip:
+      state: present
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: `organization` parameter is required

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    api = Scaleway(module=module)
+    changed, summary = state_strategy[module.params["state"]](api=api, wished_ip=wished_ip)
+    module.exit_json(changed=changed, scaleway_ip=summary)
+
+
+def main():
+    argument_spec = scaleway_argument_spec()
+    argument_spec.update(dict(
+        state=dict(default='present', choices=['absent', 'present']),
+        organization=dict(),
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: `organization` must be required

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        if api.module.check_mode:
+            return changed, {"status": "An IP would be created."}
+
+        # Create IP
+        creation_response = api.post('/ips',
+                                     data=payload_from_wished_ip(wished_ip))
+
+        if not creation_response.ok:
+            msg = 'Error during ip creation: (%s) %s' % (
+                creation_response.status_code, creation_response.body)
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py]: `creation_response.body` is empty I propose to use:
```
"%s: '%s' (%s)" % (creation_response.info['msg'], creation_response.json['message'], creation_response.json)
```

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py (hunk 1 above)
Comment (maxamillion) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: I also think this would be advantageous.

Comment (remyleone) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: Started working on it there: https://github.com/ansible/ansible/pull/44826/files 

Could you review and merge it?

**Comment by pilou-** [comment]
@sieben a rebase is required for this one

**Comment by remyleone** [comment]
Rebase done. I'm going to add additional tests for assigning ip to scaleway_compute

**Comment by maxamillion** [comment]
rebuild_merge

File: lib/ansible/modules/cloud/scaleway/scaleway_ip.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+      - ams1
+      - EMEA-NL-EVS
+      - par1
+      - EMEA-FR-PAR1
+
+  id:
+    description:
+    - id of the Scaleway IP (UUID)
+
+  server:
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_ip.py:56]: Didn't we agree that:
- this module should not handle that
- scaleway_compute should be used instead
?

Comment (remyleone) [reply lib/ansible/modules/cloud/scaleway/scaleway_ip.py:56]: Indeed but I guess we can fix it in another PR.

**Pull request merged**